EIPS/eip-2315.md

```diff
@@ -0,0 +1,78 @@
+---
+eip: 2315
+title: Simple Subroutines for the EVM
+status: Draft
+type: Standards Track
+category: Core
+author: Greg Colvin
+discussions-to: https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm/3941
+created: 2019-10-17
+---
+## Abstract
+
+This proposal introduces two opcodes to support subroutines:  `JUMPSUB` and `RETSUB`.
+
+## Motivation
+
+The EVM does not provide subroutines as a primitive.  Instead, calls must be synthesized by fetching and pushing the current program counter on the data stack and jumping to the subroutine address; returns must be synthesized by getting the return address to the top of stack and jumping back to it.
+
+## Specification
+
+##### `JUMPSUB`
+Jumps to the address on top of the stack, which must be the offset of a `JUMPDEST`.
+
+##### `RETSUB`
+Returns to the most recently executed `JUMPSUB` instruction.
+
+## Rationale
+
+This is the smallest possible change that provides native subroutines without breaking backwards compatibility.
+
+## Backwards Compatibility
+
+These changes do not affect the semantics of existing EVM code.
+
+## Test Cases
+```
+step op       stack
+0    PUSH1 3  []
+1    JUMPSUB  [3]
+4    STOP
+2    JUMPDEST []
+3    RETSUB   []
+```
+## Implementations
+
+No clients have implemented this proposal as of yet.
+
+The new operators proposed here are implemented by the following pseudocode, which adds cases for `JUMPSUB` and `RETSUB` to a simple loop-and-switch interpreter.
+```
+bytecode[code_size]
+data_stack[1024]
+return_stack[1024]
+while PC < code_size {
+   switch opcode = bytecode[PC] {
+   ...
+   case JUMPSUB:
+       push(return_stack, PC)
+       PC = pop(data_stack)
+   case RETSUB:
+       PC = pop(return_stack)
+   ...
+   }
+   ++PC
+}
+```
+Execution of EVM bytecode begins with one value on the return stack—the size of the bytecode. The virtual byte of 0 at this offset is the EVM `STOP` opcode, so executing a `RETSUB` with no prior `JUMPSUB` executes a `STOP`.  A `STOP` or `RETURN` ends the execution of the subroutine and the program.
+
+We suggest the cost of `JUMPSUB` should be _low_, and `RETSUB` should be _verylow_. 
+ Measurement will tell.  We suggest the following opcodes:
+```
+0xbe JUMPSUB
+0xbf RETSUB
+```
+## Security Considerations
+
+This proposal introduces no new security considerations to the EVM.
```

Comment:
It kind of does, there are a couple of new things: 

- Program flow analysis frameworks need to be update for a new type of branching, since a `RETSUB` will cause a jump to a destination which is not a `JUMPDEST`.

Reply:
Could consider introducing the requirement of `JUMPDEST` after `JUMPSUB`?

Reply:
I was considering that, and think it's a good idea.  Will simplify things that cause a long, confusing discussion above.

Reply:
I tried that, but am not so sure now that it's a good idea.  Flow analysis frameworks will already have to change to allow for JUMPSUB, which is new kind of branching.  And we still don't know statically where a RETSUB will go, except that it will for sure be the instruction after the most recent JUMPSUB.

Reply:
That is, since you can't statically check the destination of a RETSUB anyway, it only figures as a basic-block terminator.  So it doesn't seem like a big enough change to waste a byte after every JUMPSUB.

Reply:
You can't statically check jump destination currently, hence the reason for JUMPDEST which statically provides potential destinations. Following that logic, how is that bad for JUMPSUB?